src/plugins/intel_cpu/src/nodes/scatter_update.cpp

```diff
@@ -263,6 +263,377 @@ static std::vector<size_t> getBlockND(const VectorDims& shape) {
+    const Reduction reduction_type = config.reduction_type;
+
+    if (axis < 0)
+        axis += updates_rank;
+
+    VectorDims squashed_indices_shape(indices_shape);
+    squashed_indices_shape[axis] = 1;
+
+    // process serially along 'axis' dimension because of data dependency brought by duplicated value in indices
+    parallel_nt(0, [&](const int ithr, const int nthr) {
+        size_t start = 0, end = 0;
+        splitter(shape_size(squashed_indices_shape), nthr, ithr, start, end);
+
+        if (!use_init_val) {
+            const auto value = reduction_neutral_value<DataType>(reduction_type);
+            for (size_t worker = start; worker < end; worker++) {
+                VectorDims indices_coord(updates_rank, 0);
+                getCoordinate(indices_coord, worker, squashed_indices_shape);
+                std::vector<size_t> data_coord(indices_coord);
+
+                for (size_t i = 0; i < index_dim_size; i++) {
+                    indices_coord[axis] = i;
+                    IndexType idxValue = indices_buf.at<IndexType, size_t>(indices_coord);
+                    if (idxValue < 0) idxValue += data_dim_size;
+                    ASSERT_DEBUG_ONLY(idxValue < data_dim_size && idxValue >= 0, "invalid index value.");
+                    data_coord[axis] = idxValue;
+                    data_buf.at<DataType, size_t>(data_coord) = value;
+                }
+            }
+        }
+
+        for (size_t worker = start; worker < end; worker++) {
+            VectorDims indices_coord(updates_rank, 0);
+            getCoordinate(indices_coord, worker, squashed_indices_shape);
+            std::vector<size_t> data_coord(indices_coord);
+
+            std::unordered_map<size_t, int64_t> mean_reduction_counters;
+
+            // inner axis loop for better performance
+            for (size_t i = 0; i < index_dim_size; i++) {
+                indices_coord[axis] = i;
+                IndexType idxValue = indices_buf.at<IndexType, size_t>(indices_coord);
+                if (idxValue < 0) idxValue += data_dim_size;
+                ASSERT_DEBUG_ONLY(idxValue < data_dim_size && idxValue >= 0, "invalid index value.");
+                data_coord[axis] = idxValue;
+                DataType& dst = data_buf.at<DataType, size_t>(data_coord);
+                DataType src = updates_buf.at<DataType, size_t>(indices_coord);
+
+                kernel_func(std::addressof(dst), &src);
+
+                mean_reduction_counters[idxValue] += 1;
+            }
+
+            for (const auto& counter : mean_reduction_counters) {
+                data_coord[axis] = counter.first;
+                DataType& dst = data_buf.at<DataType, size_t>(data_coord);
+                const auto N = counter.second + static_cast<int32_t>(use_init_val);
+                dst = static_cast<DataType>(static_cast<double>(dst) / N);
+            }
+        }
+    });
+}
+
+struct Caller : public element::NoAction<bool> {
+    using element::NoAction<bool>::visit;
+
```

Comment:
why introduce this level of complexity?

Reply:
https://github.com/openvinotoolkit/openvino/pull/23287#discussion_r1531530506

Reply:
refacted